### PR TITLE
V1

### PR DESCRIPTION
**Pull request by ettomatic**
A new non-backwards compatible version of Crimp.
- reduces the input data structure to language agnostic primitives
- works with any version of Ruby
- raise an error if custom objects get passed
- deep sort nested arrays and heashes
- better testing

More infos in the updated README

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,42 +3,141 @@
...
-answer](http://stackoverflow.com/a/6462589/3243663).
+Creates an MD5 hash of a data structure made of numbers, strings, booleans, arrays, sets or hashes.
 
 ## Installation
 
 Add this line to your application's Gemfile:
 
-    gem 'crimp'
+```
+$ gem 'crimp'
```
Comment (samfrench) [inline README.md]: We don't need the $ if this line is added to the Gemfile.

File: README.md [hunk 2]
```diff
@@ -3,42 +3,141 @@
 [![Build Status](https://travis-ci.org/BBC-News/crimp.png?branch=master)](https://travis-ci.org/BBC-News/crimp)
 [![Gem Version](https://badge.fury.io/rb/crimp.png)](http://badge.fury.io/rb/crimp)
 
-Creating an md5 hash of a number, string, array, or hash in Ruby
-
-![mighty-boosh-four-way-crimp-o](https://f.cloud.github.com/assets/180050/2148112/b44fd6fa-93de-11e3-9f9a-ad941f069b5c.gif)
-
-Shamelessly copied from [this Stack Overflow
-answer](http://stackoverflow.com/a/6462589/3243663).
+Creates an MD5 hash of a data structure made of numbers, strings, booleans, arrays, sets or hashes.
```
Comment (samfrench) [inline README.md]: I know we flatten the arrays, but this says we create an MD5 hash of arrays. Would we expect these to return the same signature?
```rb
Crimp.signature([])
=> "d41d8cd98f00b204e9800998ecf8427e"
Crimp.signature([[]])
=> "d41d8cd98f00b204e9800998ecf8427e"
```

Comment (samfrench) [reply README.md]: Not sure if an empty array should return the same as an empty hash?
```rb
Crimp.signature({})
=> "d41d8cd98f00b204e9800998ecf8427e"
```

Comment (ettomatic) [reply README.md]: Good point. This happens because the idea is to reduce the content to its primitives. so yes, in this case, an empty hash == an empty array...

File: lib/crimp.rb [hunk 3]
```diff
@@ -1,74 +1,43 @@
...
+      when Numeric
+        [obj, 'n']
+      when TrueClass, FalseClass
+        [obj, 'b']
+      when NilClass
+        [nil, '_']
+      when Enumerable
+        process_collection(obj)
+      else
+        raise ArgumentError
```
Comment (samfrench) [inline lib/crimp.rb]: Should we say we don't support the type being parsed?

Comment (ettomatic) [reply lib/crimp.rb]: I'm already on it

File: README.md (hunk 2 above)
Comment (ettomatic) [reply README.md]: I can easily add a 'byte' to indicate the data structure type:
```ruby
Crimp.to_a([123, 'abc'])
=>  [[[123, "N"], ["abc", "S"]], "A"]
```

Comment (ettomatic) [reply README.md]:
```ruby
Crimp.to_a({a: 'b'})
=>  [[[[["a", "S"], ["b", "S"]], "A"]], "H"]
```

File: lib/crimp.rb [hunk 4]
```diff
@@ -1,74 +1,43 @@
...
-    case obj
-    when Array
-      parse_array obj
-    when Hash
-      parse_hash obj
-    when String
-      obj
-    else
-      to_string obj
+require 'digest'
```
Comment (samfrench) [inline lib/crimp.rb]: Should we require just the md5 class?
```rb
require 'digest/md5'
```

File: spec/crimp_spec.rb [hunk 5]
```diff
@@ -1,99 +1,208 @@
...
+    expect(Crimp.to_a(nil)).to eq([nil, '_'])
+  end
+end
+
+describe 'Booleans' do
+  it 'handles falsey values' do
+    expect(Crimp.to_a(false)).to eq([false, 'b'])
+  end
+
+  it 'handles thruty values' do
```
Comment (samfrench) [inline spec/crimp_spec.rb]: typo thruty

File: spec/crimp_spec.rb [hunk 6]
```diff
@@ -1,99 +1,208 @@
...
+  end
+
+  it 'sorts hashes by key and then sorts the resulting pair of tuples' do
+    expected = [
+      [
+        ['a', 's'],
+        ['b', 's']
+      ],
+      [
+        [1,   'n'],
```
Comment (samfrench) [inline spec/crimp_spec.rb]: The spacing seems a bit out.

File: README.md (hunk 2 above)
Comment (ettomatic) [reply README.md]: I've added a commit to this PR to address this issue, if is fine I'll update the README and add a couple of tests for edge cases

File: spec/crimp_spec.rb [hunk 7]
```diff
@@ -106,50 +112,75 @@
...
+    hsh2 = { a: 2, b: 1 }
 
     expect(Crimp.signature(hsh1)).to_not eq(Crimp.signature(hsh2))
   end
 end
 
 describe 'Sets' do
   it 'handles sets' do
-    expect(Crimp.to_a(Set.new([1, 2]))).to eq([[1, 'n'],[2, 'n']])
+    expect(Crimp.to_a(Set.new([1, 2]))).to eq([[[1, 'N'], [2, 'N']], 'S'])
```
Comment (samfrench) [inline spec/crimp_spec.rb]: As you pointed out S is used for strings and sets:

```rb
Crimp.signature('a') == Crimp.signature(Set['a'])
=> true
```

File: spec/crimp_spec.rb [hunk 8]
```diff
@@ -1,99 +1,269 @@
...
+    hsh1 = { a: 1, b: 2 }
+    hsh2 = { a: 2, b: 1 }
+
+    expect(Crimp.signature(hsh1)).to_not eq(Crimp.signature(hsh2))
+  end
+end
+
+describe 'Sets' do
+  it 'handles sets' do
+    expect(Crimp.to_a(Set.new([1, 2]))).to eq([[[1, 'N'], [2, 'N']], 'S'])
```
Comment (samfrench) [inline spec/crimp_spec.rb]: As you pointed out S is used for strings and sets:

```rb
Crimp.signature('a') == Crimp.signature(Set['a'])
=> true
```

File: README.md (hunk 2 above)
Comment (samfrench) [reply README.md]: Set will need to be removed from the readme now.

File: README.md [hunk 9]
```diff
@@ -3,42 +3,141 @@
...
+| nil     |  `_` |
+
+You can verify it using the '#to_a' method:
+
+```ruby
+Crimp.to_a({ a: { b: 'c' } })
+=> [[["a", "s"], [[["b", "s"], ["c", "s"]]]]]
+```
+
+Please note the Arrays, Sets and Hash keys are sorted before signing.
```
Comment (samfrench) [inline README.md]: Also from here.

File: crimp.gemspec [hunk 10]
```diff
@@ -1,35 +1,23 @@
...
-
-                         All credit should go to http://stackoverflow.com/users/394282/luke
-                       EOS
-  spec.homepage      = ""
-  spec.license       = "MIT"
+  spec.name          = 'Crimp'
+  spec.version       = '1.0.0'
+  spec.authors       = ['BBC News']
+  spec.email         = ['D&ENewsFrameworksTeam@bbc.co.uk']
+  spec.summary       = 'Creates an md5 hash from numbers, strings, arrays, sets or hashes.'
```
Comment (samfrench) [inline crimp.gemspec]: And here.

File: README.md (hunk 9 above)
Comment (ettomatic) [reply README.md]: I'm working on the documentation

File: crimp.gemspec (hunk 10 above)
Comment (ettomatic) [reply crimp.gemspec]: I'm working on the documentation

File: README.md [hunk 11]
```diff
@@ -3,45 +3,163 @@
...
+```
 
-# => [\"aSymbol=>[\\\"bSymbol=>b\\\", \\\"cSymbol=>c\\\"]Array\",\"dSymbol=>d\"]Array"
+## Multiplatform design
 
-Crimp.signature({:a => {:b => 'b', :c => 'c'}, :d => 'd'})
+At the BBC we use Crimp to build keys for database and cache entries.
 
-# => "68d07febc4f47f56fa6ef5de063a77b1"
+If you want to build a similiar library with your language of choice you should be able to follow the simple specifications defined in `spec/crimp_spec.rb`. Using these simple rules you will produce a string ready to be MD5 signed.
```
Comment (samfrench) [inline README.md]: typo similiar

File: README.md [hunk 12]
```diff
@@ -3,45 +3,171 @@
...
+Also Sets get transformed to Arrays:
+
+``` ruby
+Crimp.signature(Set.new(['a', 'b'])) == Crimp.signature(['a', 'b'])
+```
+
+Also remember that collections gets sorted so:
+
+``` ruby
+Crimp.signature([2, 1]) == Crimp.Signature([1, 2])
```
Comment (samfrench) [inline README.md]: capital Signature

File: README.md [hunk 13]
```diff
@@ -3,45 +3,171 @@
...
+Notice how Crimp marks the collection as Hash (`H`) and then transforms the tuple of key/values to an Array (`A`).
+
+Here's an example with nested hashes:
 
+```ruby
+Crimp.to_a({ a: { b: 'c' } })
+=> [[[[["a", "S"], [[[[["b", "S"], ["c", "S"]], "A"]], "H"]], "A"]], "H"]
 ```
 
+Before signing Crimp trasform the collection of nested array to a string.
```
Comment (samfrench) [inline README.md]: Typo transform

File: spec/crimp_spec.rb [hunk 14]
```diff
@@ -1,99 +1,269 @@
...
+    expect(Crimp.to_a(nil)).to eq([nil, '_'])
+  end
+end
+
+describe 'Booleans' do
+  it 'handles falsey values' do
+    expect(Crimp.to_a(false)).to eq([false, 'B'])
+  end
+
+  it 'handles thruty values' do
```
Comment (samfrench) [inline spec/crimp_spec.rb]: typo thruty

File: README.md [hunk 15]
```diff
@@ -3,45 +3,171 @@
 [![Build Status](https://travis-ci.org/BBC-News/crimp.png?branch=master)](https://travis-ci.org/BBC-News/crimp)
 [![Gem Version](https://badge.fury.io/rb/crimp.png)](http://badge.fury.io/rb/crimp)
 
-Creating an md5 hash of a number, string, array, or hash in Ruby
-
-![mighty-boosh-four-way-crimp-o](https://f.cloud.github.com/assets/180050/2148112/b44fd6fa-93de-11e3-9f9a-ad941f069b5c.gif)
```
Comment (woodyblah) [inline README.md]: (!https://media1.tenor.com/images/a4707fd9e9620da090345277a7c0631c/tenor.gif?itemid=9134500)

Comment (ettomatic) [reply README.md]: chimp or crimp?

File: spec/acceptance_data.yml [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+  string: _1N1S3NA
+  signature: 518e7bb17674f6acbb296845862a152d
+-
+  desc: verify hash like data structures
+  json: {"a": 1}
+  string: 1NaSAH
+  signature: 8cb44d69badda0f34b0bab6bb3e7fdbf
+-
+  desc: verify nested hash
+  json: {"a": {"c": nil, "2": 2 }}
```
Comment (coderkind) [inline spec/acceptance_data.yml]: We couldn't pass this exact data structure into a `signature` function in other languages such as JavaScript, as `nil is not defined`. Would have to be `'nil'` or `null` I expect (but you'd get a different MD5 hash produced as a result).

Comment (ettomatic) [reply spec/acceptance_data.yml]: oops thus should be `null` as it is in line :14 
👍

Comment (coderkind) [reply spec/acceptance_data.yml]: Do you need a `number` data type example in here?

File: spec/crimp_spec.rb [hunk 17]
```diff
@@ -1,99 +1,269 @@
...
+    expect(Crimp.to_a(Set.new([1, 2]))).to eq([[[1, 'N'], [2, 'N']], 'A'])
+  end
+
+  it 'sorts sets as arrays' do
+    expect(Crimp.to_a(Set.new([2, 1]))).to eq([[[1, 'N'], [2, 'N']], 'A'])
+  end
+end
+
+describe 'nested data structures' do
+  it 'handles an hash with nested arrays' do
```
Comment (samfrench) [inline spec/crimp_spec.rb]: with nested items? One of them is a hash.

File: spec/acceptance_data.yml [hunk 18]
```diff
@@ -0,0 +1,41 @@
+---
+-
+  desc: verify String handling
+  json: abc
+  string: abcS
+  signature: c4449120506d97975c67be69719a78e2
+-
```
[the remark this answers is not in the document]
Comment (ettomatic) [reply spec/acceptance_data.yml]: added!

File: lib/crimp.rb [hunk 19]
```diff
@@ -1,74 +1,52 @@
...
-  def self.legacy_class(obj)
-    return obj.class unless obj.is_a?(Numeric)
-    return 'Float'   if obj.is_a?(Float)
-    return 'Bignum'  if obj.bignum?
-    'Fixnum'
+    def coerce(obj)
+      case obj
+      when Symbol then obj.to_s
+      when Set    then obj.to_a
+      else obj
```
Comment (samfrench) [inline lib/crimp.rb]: Won't this one return nil?

File: spec/acceptance_data.yml [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+  desc: verify nested Arrays
+  json: ["a", 1, ["b", "2"]]
+  string: 1N2SbSAaSA
+  signature: 3aaa58da4841eaeb41d3726d2c6fd875
+-
+  desc: verify nested Arrays
+  json: [["b", "2"], "a", 1]
+  string: 1N2SbSAaSA
+  signature: 3aaa58da4841eaeb41d3726d2c6fd875
+
```
Comment (samfrench) [inline spec/acceptance_data.yml]: Why the extra line?

File: spec/acceptance_data.yml [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+  json: {"a": 1}
+  string: 1NaSAH
+  signature: 8cb44d69badda0f34b0bab6bb3e7fdbf
+-
+  desc: verify nested hash
+  json: {"a": {"c": null, "2": 2 }}
+  string: aS2S2NA_cSAHAH
+  signature: bff3538075e4007c7679a7ba0d0a5f30
+-
+  desc: verify nil values
```
Comment (samfrench) [inline spec/acceptance_data.yml]: Should this be null if this is language independent?

File: spec/crimp_spec.rb [hunk 22]
```diff
@@ -1,99 +1,299 @@
...
+    expect(Crimp.signature(arr1)).to_not eq(Crimp.signature(arr2))
+  end
+
+  it 'sorts an array with mixed strings and symbols' do
+    expect(Crimp.notation(["b", :a, "c"])).to eq 'aSbScSA'
+  end
+end
+
+describe 'Nested Arrays' do
+  it 'sorts arrays with a single nested array ' do
```
Comment (samfrench) [inline spec/crimp_spec.rb]: Minor space before closing quote.

**Comment by samfrench** [comment]
Is it worth linking to our rubocop config repo in case anyone wanted to contribute to this repo?

File: spec/crimp_spec.rb [hunk 23]
```diff
@@ -1,99 +1,299 @@
...
+
+describe '.notation' do
+  it 'returns a string representation of the passed data' do
+    expect(Crimp.notation([123, 'abc'])).to eq('123NabcSA')
+  end
+end
+
+describe '.reduce' do
+  it 'returns an array of tuples representing the value and the type' do
+    expect(Crimp.reduce([123, 'abc'])).to eq([[[123, 'N'], ['abc', 'S']], 'A'])
```
Comment (samfrench) [inline spec/crimp_spec.rb]: Not sure this is reducing the input as it's adding notation values and now seems larger.

Comment (samfrench) [reply spec/crimp_spec.rb]: categorise ?

Comment (ettomatic) [reply spec/crimp_spec.rb]: thinking the same... it's not reducing. better name?

Comment (samfrench) [reply spec/crimp_spec.rb]: categorising, labelling, classifying, tagging or indexing.

Comment (ettomatic) [reply spec/crimp_spec.rb]: `Crimp.to_tuple`,'`Crimp.build`,`Crimp.to_sequence`,`Crimp.concat`,`Crimp.stack`

Comment (woodyblah) [reply spec/crimp_spec.rb]: I like categorise, tuple, or sequence. Or perhaps `Crimp.scaffold` or `Crimp.structure`

Comment (samfrench) [reply spec/crimp_spec.rb]: annotate? label?

Comment (ettomatic) [reply spec/crimp_spec.rb]: annotate

File: spec/crimp_spec.rb [hunk 24]
```diff
@@ -1,99 +1,299 @@
...
+describe 'Numbers' do
+  it 'handles integers' do
+    expect(Crimp.reduce(1)).to eq([1, 'N'])
+  end
+
+  it 'handles floats' do
+    expect(Crimp.reduce(3.14)).to eq([3.14, 'N'])
+  end
+
+  it 'handles bignums' do
```
[the remark this answers is not in the document]
Comment (ettomatic) [reply spec/crimp_spec.rb]: I can't see any problem

**Comment by ettomatic** [comment]
re: rubocop repo: I don't think it's necessary. Ut's such a small library..
